cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -993,21 +994,43 @@ std::shared_ptr<std::unordered_map<std::pair<StreamId, VersionId>, AtomKey>> Loc
         specific_versions = std::make_shared<std::unordered_map<std::pair<StreamId, VersionId>, AtomKey>>();
         latest_versions = batch_get_latest_version(store(), version_map(), stream_ids, false);
     }
-    for(const auto& latest_version : *latest_versions)
+
+    for(const auto& latest_version : *latest_versions) {
+        (*latest_versions_ids)[latest_version.first] = latest_version.second.version_id();
         specific_versions->try_emplace(std::make_pair(latest_version.first, latest_version.second.version_id()), latest_version.second);
+    }
 
-    return specific_versions;
+    return std::make_pair(specific_versions, latest_versions_ids);
 }
 
 std::vector<std::pair<VariantKey, std::optional<google::protobuf::Any>>> LocalVersionedEngine::batch_read_metadata_internal(
     const std::vector<StreamId>& stream_ids,
     const std::vector<VersionQuery>& version_queries
     ) {
-    auto stream_index_map = get_stream_index_map(stream_ids, version_queries);
+    auto [stream_index_map, latest_versions_ids] = get_stream_index_map(stream_ids, version_queries);
     std::vector<folly::Future<std::pair<VariantKey, std::optional<google::protobuf::Any>>>> fut_vec;
-    for (const auto &[_, key] : *stream_index_map)
-        fut_vec.push_back(store()->read_metadata(key));
-
+    for(const auto& stream_id : folly::enumerate(stream_ids)) {
+        if(!version_queries.empty()) {
+            const auto& query = version_queries[stream_id.index].content_;
+            if(std::holds_alternative<SpecificVersionQuery>(query)){
+                auto specific_version_map_key = std::make_pair(*stream_id, std::get<SpecificVersionQuery>(query).version_id_);
+                auto specific_version_it = stream_index_map->find(specific_version_map_key);
+                if (specific_version_it != stream_index_map->end()){
+                    fut_vec.push_back(store()->read_metadata(specific_version_it->second));
+                    continue;
+                }
+            }
+        }
+        auto last_version_symbol_it = latest_versions_ids->find(*stream_id);
+        if(last_version_symbol_it != latest_versions_ids->end()) {
+            auto last_version_map_key = std::make_pair(*stream_id, last_version_symbol_it->second);
+            auto last_version_it = stream_index_map->find(last_version_map_key);
+            util::check(last_version_it != stream_index_map->end(), "If we didn't specify or have not found a specific version, we must be able to index this map with the last one");
+            fut_vec.push_back(store()->read_metadata(last_version_it->second));
+        }else{
```

Comment:
When do we hit this?

Reply:
Basically we need to track the cases where we do not specify the version (either because the vector of queries is empty or because that specific entry does not contain a specific version id). As commented above is more clear now